### DISCUSSION TITLE
How do I get rid of UI Flicker?

### DISCUSSION THREAD

**Discussion by rexwreyes** [cloudscape-design/components#2100] [Q&A] [answered]
I am using the AppLayout and TopNavigation components in an NextJS Amplify App.  When I click on any (they are all internal) link in the TopNavigation to goto a new page it looks like the css unloads and reloads.  This happens on every page.  Here is an example of it loading...
![loading](https://github.com/cloudscape-design/components/assets/38086223/84d320f6-f495-449f-a480-347785016645)

and loaded...
![loaded](https://github.com/cloudscape-design/components/assets/38086223/864f66e6-b2f2-4f16-8a8a-fc495798b2c1)

I notice on the AWS site (the sections that use cloudscape) that when you click on a link you get an overlay while things load and you don't get this flicker.  how can I achieve this?

**Comment by connorlanigan** [comment, marked as answer]
It looks like your application might be performing a full page load when you navigate to a new page, instead of performing a client-side navigation only. You can specify handlers on the TopNavigation component that are called when a link in the TopNavigation is clicked:
- the object in the `identity` prop takes an `onFollow` field
- the objects in the `utilities` prop take an `onFollow` (for buttons) or `onItemFollow` (for menu dropdowns) field

In those handlers, you can then cancel the browser's full-page navigation and perform a client-side navigation using your framework's routing functionality. In NextJS, you should be able to use the `useRouter` hook (see [the NextJS documentation](https://nextjs.org/docs/app/api-reference/functions/use-router)) for that.
Something like this:
```tsx
const router = useRouter();

return <TopNavigation
          identity={{
            href: '/',
            onFollow: e => {
              e.preventDefault();
              router.push('/');
            },
          }}
        />
```
